### ISSUE TITLE
Upload data failed

### ISSUE THREAD

**Issue by xiyue-55** [aws-solutions/amazon-marketing-cloud-uploader-from-aws#264]
Hello,

I tried to upload the data with Uploader, but always got AWS Glue Job failed in **Step 6 - [Monitor uploads]** 
I've well configured all the steps and there wasn't any error showed in the configuration steps.
I checked the logs in CloudWatch, but don't understand what could the error be from my side.

Do you have any idea on the problem?
Thanks for your help!

Xiyue

- [ ] [ v3.0.0]

```yaml
Mappings:
  Application:
    Solution:
      Id: "SO0222"
      Name: "amcufa"
      Version: "v3.0.0"
```

- [ ] Region: eu-west-1
- [ ] [CloudWatch Logs ] 
2024-07-08 14:50:33,217 ERROR [main] glue.ProcessLauncher (Logging.scala:logError(73)): Error from Python:Traceback (most recent call last):
  File "/usr/lib64/python3.7/argparse.py", line 1787, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib64/python3.7/argparse.py", line 1993, in _parse_known_args
    start_index = consume_optional(start_index)
  File "/usr/lib64/python3.7/argparse.py", line 1923, in consume_optional
    arg_count = match_argument(action, selected_patterns)
  File "/usr/lib64/python3.7/argparse.py", line 2088, in _match_argument
    raise ArgumentError(action, msg)
argparse.ArgumentError: argument --country_code: expected one argument
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/tmp/amc_transformations.py", line 70, in check_params
    args = getResolvedOptions(sys.argv, required)
  File "/opt/amazon/lib/python3.6/site-packages/awsglue/utils.py", line 119, in getResolvedOptions
    parsed, extra = parser.parse_known_args(args[1:])
  File "/usr/lib64/python3.7/argparse.py", line 1794, in parse_known_args
    self.error(str(err))
  File "/opt/amazon/lib/python3.6/site-packages/awsglue/utils.py", line 76, in error
    raise GlueArgumentError(msg)
awsglue.utils.GlueArgumentError: argument --country_code: expected one argument
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/tmp/amc_transformations.py", line 124, in <module>
    params = check_params(required=REQUIRED_PARAMS, optional=OPTIONAL_PARAMS)
  File "/tmp/amc_transformations.py", line 73, in check_params
    sys.exit(1)
SystemExit: 1

**Comment by xiyue-55** [comment]
Update: new errors

2024-07-09 13:50:16,915 ERROR [main] glue.ProcessLauncher (Logging.scala:logError(73)): Error from Python:Traceback (most recent call last):
  File "/tmp/amc_transformations.py", line 70, in check_params
    args = getResolvedOptions(sys.argv, required)
  File "/opt/amazon/lib/python3.6/site-packages/awsglue/utils.py", line 119, in getResolvedOptions
    parsed, extra = parser.parse_known_args(args[1:])
  File "/usr/lib64/python3.7/argparse.py", line 1787, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib64/python3.7/argparse.py", line 2022, in _parse_known_args
    ', '.join(required_actions))
  File "/opt/amazon/lib/python3.6/site-packages/awsglue/utils.py", line 76, in error
    raise GlueArgumentError(msg)
awsglue.utils.GlueArgumentError: the following arguments are required: --destination_endpoints

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/amc_transformations.py", line 124, in <module>
    params = check_params(required=REQUIRED_PARAMS, optional=OPTIONAL_PARAMS)
  File "/tmp/amc_transformations.py", line 73, in check_params
    sys.exit(1)
SystemExit: 1

**Comment by xiyue-55** [comment]
Update:
I redeployed the cloudformation stack. The errors about the destination_endpoints is fixed. Actually it was because cloudFormation stack update don't well replaced the Glue Job library and notebook with the new version.

But I met a new bug:
The Glue Job status shows succeeded, but I got failed in Uploads status like the image shows below
![image](https://github.com/aws-solutions/amazon-marketing-cloud-uploader-from-aws/assets/163162007/ed6d0bbf-0b7a-4aba-ba5f-8476d328b2c1)

In the logs of glue job I got this:
Writing 295028 rows to s3://amc-uploader-v3-etl-artifacts-4bolkg/amc/ns_one_shot_unified_profiles/ADDITIVE/CSV/**null**/amcv7agsrv7|eu-west-1_pcplZllWv/fileName.csv-0.gz

Do you think it's the null in the S3 URI caused the problem?

**Comment by ianwow** [comment]
@xiyue-55 Thanks for submitting this issue. I've a couple follow-up questions:

- What country code did you specify when you performed the upload?
- Did you perform the upload using the web application or programmatically by, for example, a custom script?
- How did you install the this solution? Did you upgrade a previous stack or create a new stack with the latest template (https://solutions-reference.s3.amazonaws.com/amazon-marketing-cloud-uploader-from-aws/latest/amazon-marketing-cloud-uploader-from-aws.template)?

**Comment by xiyue-55** [comment]
Hi Ian,
- The country code is IT,
- I did use uploader web application
- I have redeployed the whole solution, the countryCode bug is not resolved. As I mentioned in my post, I've upgraded Uploader in AWS CloudFormation by updating the stack => replace existing template. But a lot of code is not updated. So in the future, should I always create a new stack when there's a new release?
The insatnceId problem is resolved as we don't correctly configured the Data Upload AWS Account in the settings.

Thanks for your help!

**Comment by ianwow** [comment]
Yes, you should always create a new stack when there's a new release because some versions do not support the upgrade option. 

I'm glad to hear you were able to resolve the InstanceId problem.

Regarding the null country code, I'm having trouble reproducing this issue with the following procedure. Does this procedure reflect what you did when you saw the Glue job fail with an error about the null country code?

1. Deploy new stack with latest AMC Uploader template
2. Log into the web application, setup the AMC connections on the Settings page.
3. Select a file on Step 1
4. Select a destination instance on Step 2
5. Select "Add to Existing Dataset" on Step 3, and select "OVERLAP_REPLACE", then select a country code, and CSV/JSON file format. Click Next
6. Click Next on Step 4 - Define Columns.
7. Click Submit on Step 5 - Confirm Details.

**Comment by xiyue-55** [comment]
Hi,
Except the point 5, on the Step 3 I choosed ADDICTIVE, the others were exactly like you describe.

**Comment by ianwow** [comment]
When you saw the error in the Glue job, did you specify any PII fields?

Country code will become null when a user uploads a dataset that does not contain any PII fields. This is a bug that will be fixed in the next release. I think you encountered this bug.

The log you specified above does, in fact, include PII fields, but it also says OVERLAP_REPLACE. Since you said you chose ADDITIVE, I think that the above log is not from the same transaction that produced the Glue job error.

**Comment by xiyue-55** [comment]
Hi,
In fact I have tested both ADDITIVE and OVERLAP_REPLACE, the 2 has the same bug. But all my datasets contain PII. The bug appears only when I add data into existing dataset. When importing data with a new created dataet, there's no problem.

**Comment by ianwow** [comment]
It sounds like your Glue jobs always fails because it's missing the country_code argument when you try to upload to an existing dataset. I have not been able to reproduce that problem. In fact, country code is supposed to be null when uploading to an existing dataset.

To troubleshoot this further, we may need to access your Ads account. Please contact your Ads account manager or amc-support to open a ticket with our developer team.

**Comment by xiyue-55** [comment]
Hello Ian,

I don't see why the bug is from the Amazon Ads accoun. And the error is raised at File "/opt/amazon/lib/python3.6/site-packages/awsglue/utils.py", line 119, in getResolvedOptions. The function **getResolvedOptions** read only the arguments passed into a gluejob, which is from the endpoint start_amc_transformation in APIHandler lambda function. 
The country_code is optional in the endpoint start_amc_transformation, but then the error is not treated in the Glue job.

**Comment by ianwow** [comment]
I am going to publish a patch release that includes the following change, later this week:
https://github.com/aws-solutions/amazon-marketing-cloud-uploader-from-aws/pull/265

That change resolves a defect which may be related to the problem you encountered. However, because I haven't been able to reproduce the error you found, I am not sure. If you setup a ticket with your Ads account manager, then I should be able to access your AMC Uploader installation to verify. I say "Ads account manager" because they can give me access to your AWS account. I agree, the error you saw is in the Glue job from this solution, rather than your Amazon Ads account.

**Comment by xiyue-55** [comment]
Thanks alot Ian! Hoping that the new release will fix my problem. Otherwise, I'll create a ticket with Ads account manager to find a further solution with you
Xiyue

**Comment by ianwow** [comment]
@xiyue-55 We published v3.0.2 today. I think it might solve the defect that you encountered.
